### PR TITLE
[ML] BuildKite nightly debug builds

### PR DESCRIPTION
**Pull request by edsavage**
The nightly debug build runs in a separate pipeline to the PR builds and is triggered on a schedule.
It executes builds with full debug and assertions enabled in order to pick up any compilation errors as early as possible.

### CODE REVIEW

**Review by droberts195 — COMMENTED** [review]
> The nightly debug build runs in a separate pipeline to the PR builds.

I think this PR highlights a design issue with the pipelines.

Basically we have 3 types of build:

1. PR builds
2. Release branch builds
3. Debug builds

At the moment there is one pipeline that does PR builds _and_ release branch builds, and another pipeline that does debug builds.

I think we should either have one pipeline for each of the 3 types of build, or 1 pipeline that does all 3 with variables to control the differences. It seems messy to have a mix of the two approaches.

At the moment the debug build is only building for linux-x86_64, but eventually we want it to build all the platforms with debug information.

**Comment by edsavage** [comment]
> I think we should either have one pipeline for each of the 3 types of build, or 1 pipeline that does all 3 with variables to control the differences. It seems messy to have a mix of the two approaches.

Yes, repeating the steps in two physically different pipeline files is an interim solution only, as it simplifies development and debugging. 

The longer term solution is to maximize code re-use, taking advantage of BuildKite's dynamic pipeline support. Effectively then we would have a single pipeline which is auto generated and then uploaded with `buildkite-agent pipeline upload`

File: .buildkite/pipelines/build_linux.json.py [hunk 1]
```diff
@@ -93,38 +93,40 @@ def main(args):
             ],
         })
 
-    if os.environ.get("BUILDKITE_PULL_REQUEST", "false") != "false":
-        # We also always cross compile for aarch64 with full debug and assertions
-        # enabled for PR builds only. This is to detect any compilation errors
-        # as early as possible.
```
[the remark this answers is not in the document]
Comment (edsavage) [reply .buildkite/pipelines/build_linux.json.py:99]: My misunderstanding, yes this change has added more work for non-PR builds.

**Comment by edsavage** [comment]
buildkite build this

File: .buildkite/pipelines/build_linux.json.py [hunk 2]
```diff
@@ -93,29 +93,33 @@ def main(args):
...
+    # Never cross-compile for linux-aarch64 in the nightly debug build.
```
[the remark this answers is not in the document]
Comment (edsavage) [reply .buildkite/pipelines/build_linux.json.py:96]: At the moment both native and cross-compilation build are triggered for branch builds. I'll push a change making that just for native builds.

**Comment by edsavage** [comment]
buildkite build this

**Comment by edsavage** [comment]
buildkite build this

File: .buildkite/pipelines/build_linux.json.py [hunk 3]
```diff
@@ -93,20 +93,25 @@ def main(args):
...
             "agents": {
               "cpu": "6",
               "ephemeralStorage": "20G",
               "memory": "64G",
               "image": "docker.elastic.co/ml-dev/ml-linux-aarch64-cross-build:10"
             },
             "commands": [
+              f'if [[ "{args.snapshot}" != "None" ]]; then export BUILD_SNAPSHOT={args.snapshot}; fi',
+              f'if [[ "{args.version_qualifier}" != "None" ]]; then export VERSION_QUALIFIER={args.version_qualifier}; fi',
+              "env",
```
Comment (droberts195) [inline .buildkite/pipelines/build_linux.json.py:114]: Is this just temporary? Printing the whole environment is a recipe for leaking secrets, so this shouldn't stay here long term.

**Review by droberts195 — APPROVED** [review]
LGTM

If you want to keep the `env` command in for a day or two for debugging that's fine, but please remove it soon.

File: .buildkite/pipelines/build_linux.json.py (hunk 3 above)
Comment (edsavage) [reply .buildkite/pipelines/build_linux.json.py:114]: > 

Yes, these can now be removed.

**Pull request merged**